### ISSUE TITLE
[Support] How to get enabled/disabled status on an Active Directory user

### ISSUE THREAD

**Issue by nspaul** [DirectoryTree/LdapRecord#493]
**Environment:**
 - LDAP Server Type: ActiveDirectory
 - PHP Version: 8.1.9
 - Laravel Version: 9.26.1

(I'm Happy to delete this and move it to the LdapRecord-Laravel repo if necessary)

I have a Laravel app where I am tracking several attributes on our users, and I have the need to check if a User is enabled or disabled. I feel like there is a simple solution to this, but I'm running into issues with it. Basically, I have a User object and I want to be able to check this status with a `->isEnabled()` method. Is this possible?

For more context, I'm getting the user like so:
`$ldapuser = User::in($baseDN)->where('samaccountname','=',$username)->first()`

This user object is an instance of ` LdapRecord\Models\ActiveDirectory\User`

Then I want to perform certain actions if the user is enabled versus disable, so I'm doing something like this:
`if ($ldapuser->isEnabled()) {
    // do stuff
}`

When I do this, though, I'm getting this Exception:
`Call to undefined method LdapRecord\Query\Model\ActiveDirectoryBuilder::isEnabled()`

Is this possible to Eloquently get the enabled/disabled status this way? Or am I just going about it the wrong way?

Thanks!

**Comment by stevebauman** [comment]
Hey @nspaul!

Yes it's totally possible. You'll have to use the user's "User Account Control" value, along with the LdapRecord `AccountControl` object to determine if a user is enabled:

https://ldaprecord.com/docs/core/v2/active-directory/users/#user-account-control

It's not identified well in the documentation and not found via the search input. I'll update this! 👍 

Here's how you can do this:

```php
use LdapRecord\Models\User;
use LdapRecord\Models\Attributes\AccountControl;

$ldapuser = User::in($baseDN)->where('samaccountname','=',$username)->first();

$accountControl = new AccountControl(
    $ldapUser->getFirstAttribute('userAccountControl')
);

if ($accountControl->has(AccountControl::ACCOUNTDISABLE)) {
    // The user account is disabled.
}
```

Let me know if you encounter any issues with this! 👍

**Comment by stevebauman** [comment]
I forgot to mention, there is a built-in scope on Active Directory queries that you can use to filter on disabled user accounts:

https://github.com/DirectoryTree/LdapRecord/blob/ac8860be6f6451497df8de34dff3d08d36adbd35/src/Query/Model/ActiveDirectoryBuilder.php#L58-L68

```php
$ldapuser = User::in($baseDN)
    ->whereDisabled()
    ->where('samaccountname','=',$username)
    ->first();

if ($ldapUser) {
    // The located LDAP user is disabled.
}
```

**Comment by nspaul** [comment]
Thanks @stevebauman! This worked perfectly. I can work with it.

Side-note (and this is not me throwing any kind of shade): I hinted at this when I wrote up this Issue, but back in the Adldap2 days, I was able to do `ldapuser->isEnabled()` to pull the info. This is preferable simply because it reads better. Think "If this user is enable, then do this thing. Else, do this other thing". It would be awesome to build this functionality back into LdapRecord, but perhaps there is a technical holdup for that. In any case, your solution here works exactly like I need it to, so I'm good with that. LdapRecord is awesome... it saves me so much work, thanks!

**Comment by stevebauman** [comment]
Excellent! Glad you're up and running @nspaul.

> Side-note (and this is not me throwing any kind of shade): I hinted at this when I wrote up this Issue, but back in the Adldap2 days, I was able to do ldapuser->isEnabled() to pull the info. This is preferable simply because it reads better.

Hear you loud and clear! I agree with you. I've just added these methods into the included `ActiveDirectory\User` model in [v2.17.0](https://github.com/DirectoryTree/LdapRecord/releases/tag/v2.17.0). You can now call `isEnabled()` or `isDisabled()`:

```php
use LdapRecord\Models\ActiveDirectory\User;

$user = User::first();

if ($user->isEnabled()) {
    // ...
}

if ($user->isDisabled()) {
    // ...
}
```

Let me know if you have any further suggestions -- my ear is always open! 🙏
